discovery/db_discovery.go

```diff
@@ -100,31 +121,58 @@ func (dbo *DBOrchestratorPoolCache) GetOrchestrators(numOrchestrators int) ([]*n
 }
 
 func (dbo *DBOrchestratorPoolCache) Size() int {
-	return len(dbo.GetURLs())
+	count, _ := dbo.store.OrchCount(
+		&common.DBOrchFilter{
+			MaxPrice:     server.BroadcastCfg.MaxPrice(),
+			CurrentRound: dbo.rm.LastInitializedRound(),
+		},
+	)
+	return count
 }
 
-func cacheRegisteredTranscoders(node *core.LivepeerNode) error {
-	orchestrators, err := node.Eth.RegisteredTranscoders()
+func (dbo *DBOrchestratorPoolCache) cacheTranscoderPool() error {
+	orchestrators, err := dbo.lpEth.TranscoderPool()
 	if err != nil {
-		glog.Error("Could not refresh DB list of orchestrators: ", err)
-		return err
+		return fmt.Errorf("Could not refresh DB list of orchestrators: %v", err)
 	}
-	_, dbOrchErr := cacheDBOrchs(node, orchestrators)
-	if dbOrchErr != nil {
-		glog.Error("Could not refresh DB list of orchestrators: cacheDBOrchs err")
-		return err
+
+	var orchs []*common.DBOrch
+
+	for _, o := range orchestrators {
+		orchs = append(orchs, ethOrchToDBOrch(o))
 	}
 
+	dbo.cacheDBOrchs(orchs)
+
 	return nil
 }
 
-func cacheDBOrchs(node *core.LivepeerNode, orchs []*lpTypes.Transcoder) ([]*common.DBOrch, error) {
-	var dbOrchs []*common.DBOrch
-	if orchs == nil {
-		glog.Error("No new DB orchestrators created: no orchestrators found onchain")
-		return dbOrchs, nil
-	}
+func (dbo *DBOrchestratorPoolCache) pollOrchestratorInfo(ctx context.Context) error {
```

Comment:
The ticker in this loop will not tick until after `cacheRefreshInterval`. While `cacheTranscoderPool()` will fetch off-chain info from any Os in the on-chain pool, it will not fetch off-chain info from any Os that are active in the current round, but that are no longer in the on-chain pool. I recommend the following:

1. `cacheTranscoderPool()` just updates the DB with the on-chain pool and does no off-chain info fetching
2. `pollOrchestratorInfo()` fetches off-chain info of all active Os when it is called *before* it starts the polling loop so that we can make sure that we fetch the off-chain info of all active Os when `DBOrchestratorPoolCache` is created

Reply:
Ah right I didn't consider the O's that are added through backfilling. 

I would prefer option 1 , but would you be okay with duplicating `parseURI` logic here then to prevent inserting O's with a bad URL? We don't actually need the parsed URI, but we need to check whether the `ServiceURI` fetched on-chain is a valid URL. If `parseURI` returns an error we can skip insertion into the DB for that O.

Reply:
We would want to do both 1 and 2 - 1 makes sure that our DB has an up-to-date view of the active set and 2 makes sure that given an up-to-date view of the active set that we immediately fetch off-chain info for each active O when we start up.

> but would you be okay with duplicating parseURI logic here then to prevent inserting O's with a bad URL?

Calling `parseURI()` in `cacheTranscoderPool()` should work. Note that the `OrchestratorWatcher` does not check the URI before updating the DB so that should probably be updated external to this PR.

Reply:
`OrchestratorWatcher` does not update the URI so it just struck me that after abolishing backfilling from a very early block we never save the URI for orchestrators that are inserted through backfilling rather than `CacheTranscoderPool`, so those O's will never have their URI field populated and thus never be selected. 

We could make an RPC call when a transcoder becomes active to fetch its `ServiceURI` , validate it and insert it if it is valid.

Reply:
I actually meant `ServiceRegistryWatcher` in the previous post (edited to reflect that), but yeah we should validate & insert the service URI in `OrchestratorWatcher` as well.

Reply:
> We could make an RPC call when a transcoder becomes active to fetch its ServiceURI

Is this the on-chain ServiceURI or the one received from `OrchestratorInfo`[1]? If the latter, is there a possibility of a thundering herd problem? Transcoder becomes active and immediately gets hammered by Bs all trying to fetch its ServiceURI?

[1] I think we should only be caching the on-chain ServiceURI anyway?

Reply:
Actually after second thought I'm wondering if we should validate the URI at all. 

An O with a bad URI will not be selected anyway and we won't try to fetch its `OrchestratorInfo`. The part of the code that does run into trouble here however are the unit tests, because the overwritten `GetOrchestratorInfo` methods do not actually validate the URI, but the "live" method does. Perhaps we should instead add URI validation to the unit tests.

Reply:
@j0sh fetch the URI on-chain from `ServiceRegistry` , should have been clearer here, my bad!

Reply:
I'm wondering if we can't simply 'validate' a url string in the sql query to begin with?
Can we use a RegExp here to prevent selecting bad urls from the DB in the first place?

Reply:
For the purposes of this PR, I think we can just skip URI validation in `cacheTranscoderPool()`. Filtering out bad URIs via a SQL query might be a nice optimization, but I think the `parseURI()` check in `getOrchInfo()` gets the job done for now.

Reply:
fixed in d358b9f1